tokio/src/time/interval.rs

```diff
@@ -159,22 +318,52 @@ impl Interval {
     ///
     /// When this method returns `Poll::Pending`, the current task is scheduled
     /// to receive a wakeup when the instant has elapsed. Note that on multiple
-    /// calls to `poll_tick`, only the `Waker` from the `Context` passed to the
-    /// most recent call is scheduled to receive a wakeup.
+    /// calls to `poll_tick`, only the [`Waker`](std::task::Waker) from the
+    /// [`Context`] passed to the most recent call is scheduled to receive a
+    /// wakeup.
     pub fn poll_tick(&mut self, cx: &mut Context<'_>) -> Poll<Instant> {
         // Wait for the delay to be done
         ready!(Pin::new(&mut self.delay).poll(cx));
 
-        // Get the `now` by looking at the `delay` deadline
-        let now = self.delay.deadline();
+        // Get the time at which the `delay` was supposed to complete
+        let timeout = self.delay.deadline();
+
+        let now = Instant::now();
+
+        // If a tick was not missed, and thus we are being called before the
+        // next tick is due, just schedule the next tick normally, one `period`
+        // after `timeout`
+        //
+        // However, if a tick took excessively long and we are now behind,
+        // schedule the next tick according to how the user specified with
+        // `MissedTickBehavior`
+        let next = if now > timeout + Duration::from_millis(5) {
+            self.missed_tick_behavior
+                .next_timeout(timeout, now, self.period)
+        } else {
+            timeout + self.period
+        };
 
-        // The next interval value is `duration` after the one that just
-        // yielded.
-        let next = now + self.period;
         self.delay.as_mut().reset(next);
 
         // Return the current instant
-        Poll::Ready(now)
+        Poll::Ready(timeout)
+    }
+
+    /// Returns the [`MissedTickBehavior`] strategy currently being used.
+    pub fn missed_tick_behavior(&self) -> MissedTickBehavior {
+        self.missed_tick_behavior
+    }
+
+    /// Sets the [`MissedTickBehavior`] strategy that should be used.
+    pub fn set_missed_tick_behavior(&mut self, behavior: MissedTickBehavior) {
+        self.missed_tick_behavior = behavior;
+    }
+
+    /// Resets the [`MissedTickBehavior`] strategy to the default, which is
+    /// [`Burst`](MissedTickBehavior::Burst).
+    pub fn reset_missed_tick_behavior(&mut self) {
+        self.missed_tick_behavior = Default::default();
     }
```

Comment:
Is this function necessary? I included it because "why not?", but if we removed the `interval_with_missed_tick_behavior_*` functions, should we also get rid of this so that the API is simpler?

Reply:
I would remove this.